presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -35,17 +36,20 @@
     private static final int MAX_POSITIONS_EVALUATED_PER_CALL = 10000;
 
     private final OperatorContext operatorContext;
-    private final List<Type> types;
+    private final List<Type> probeTypes;
+    private final List<Type> buildTypes;
     private final ListenableFuture<? extends LookupSource> lookupSourceFuture;
     private final JoinProbeFactory joinProbeFactory;
     private final Runnable onClose;
 
-    private final PageBuilder pageBuilder;
+    private IntVector probeIndexBuilder = new IntVector();
+    private final PageBuilder buildPageBuilder;
 
     private final boolean probeOnOuterSide;
 
     private LookupSource lookupSource;
     private JoinProbe probe;
+    private boolean reachedProbeEnd = true;
```

Comment:
> Now we want to call probe's method when we reached probe's end to build the page.

I know. Suggestion here was correlated with suggestion elsewhere not to use `JoinProbe` to build the dictionary block since it's the same code in all cases.

> Also, this naming self-documents code, semantic meaning probe == null vs reachedProbeEnd self-explanatory on the right.

True. But at the cost of more state, which might go out of sync.
Naturally, this is a matter of taste. You can have clear naming _and_ less state with
```
boolean reachedProbeEnd() {
  return probe==null;
}
```